packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,192 @@
+import * as AI from '../../../index.js';
+import { ConversationMessage, FunctionResponse } from '../../../core/conversation.js';
+
+import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
+import { getEncoding } from 'js-tiktoken';
+import yaml from 'js-yaml';
+import _ from 'lodash';
+import { UseToolsProps } from '../../use-tools.js';
+import { cohereContext, reranker } from '../../../lib/cohere.js';
+
+export type LengthFunction = ((text: string) => number) | ((text: string) => Promise<number>);
+export interface RedactedFuncionResponseMetadata {
+  isRedacted: true;
+  chunks: string[];
+}
+
+const getEncoder = _.once(() => getEncoding('cl100k_base'));
+
+function openAITokenCount(text: string) {
+  return getEncoder().encode(text).length;
+}
+
+export async function TruncateByChars(
+  {
+    children,
+    maxLength,
+  }: {
+    children: string;
+    maxLength: number;
+  },
+  { render }: AI.ComponentContext
+) {
+  const stringified = await render(children);
+  if (stringified.length <= maxLength) {
+    return stringified;
+  }
+  return `${stringified.slice(0, maxLength - 3)}...`;
+}
+
+export async function LargeFunctionResponseHandler(
+  {
+    children,
+    maxLength,
+    numChunks = 4,
+    // use Cohere token counter?
+    lengthFunction = openAITokenCount,
+    ...props
+  }: AI.PropsOfComponent<typeof FunctionResponse> & {
+    maxLength: number;
+    numChunks?: number;
+    lengthFunction?: LengthFunction;
+  },
+  { render, logger, getContext }: AI.ComponentContext
+) {
+  let stringified = await render(children);
```

Comment:
For some reason I didn't see any issues with rendering tool responses that might be an object (i.e. parsed JSON vs stringified).

@petersalas since this might be relevant to your existing PR, just wanted to get you to double check this part.

https://fixie-ai.slack.com/archives/C05B6EE9JE5/p1694191388802549